src/io/flutter/run/daemon/DevToolsService.java

```diff
@@ -0,0 +1,134 @@
+import com.intellij.openapi.project.ProjectManager;
+import com.intellij.openapi.project.ProjectManagerListener;
+import com.intellij.openapi.util.io.FileUtil;
+import com.intellij.openapi.vfs.CharsetToolkit;
+import io.flutter.FlutterUtils;
+import io.flutter.android.IntelliJAndroidSdk;
+import io.flutter.bazel.Workspace;
+import io.flutter.bazel.WorkspaceCache;
+import io.flutter.sdk.FlutterSdk;
+import io.flutter.sdk.FlutterSdkUtil;
+import io.flutter.utils.MostlySilentOsProcessHandler;
+import org.jetbrains.annotations.NotNull;
+
+import static com.jetbrains.lang.dart.ide.runner.server.vmService.VmServiceWrapper.LOG;
+
+public class DevToolsService {
+  private class DevToolsServiceListener implements DaemonEvent.Listener {
+  }
+
+  @NotNull private final Project project;
+  private DaemonApi daemonApi;
+  private ProcessHandler process;
+  private DevToolsInstance devToolsInstance;
+
+  @NotNull
+  public static DevToolsService getInstance(@NotNull final Project project) {
+    return ServiceManager.getService(project, DevToolsService.class);
+  }
+
+  private DevToolsService(@NotNull final Project project) {
+    this.project = project;
+    try {
+      final GeneralCommandLine command = chooseCommand(project);
+      assert command != null;
+      this.process = new MostlySilentOsProcessHandler(command);
+      daemonApi = new DaemonApi(process);
+      daemonApi.listen(process, new DevToolsServiceListener());
+      daemonApi.devToolsServe().thenAccept((DaemonApi.DevToolsAddress address) -> {
+        if (!project.isOpen()) {
+          // We should skip starting DevTools (and doing any UI work) if the project has been closed.
+          return;
+        }
+        if (address == null) {
+          LOG.error("DevTools address was null");
+        }
+        else {
+          devToolsInstance = new DevToolsInstance(address.host, address.port);
+        }
+      });
+    }
+    catch (ExecutionException e) {
+      e.printStackTrace();
+    }
+
+    ProjectManager.getInstance().addProjectManagerListener(project, new ProjectManagerListener() {
+      @Override
+      public void projectClosing(@NotNull Project project) {
+        daemonApi.daemonShutdown();
+        if (!process.isProcessTerminated()) {
+          process.destroyProcess();
+        }
+      }
+    });
+  }
+
+  private static GeneralCommandLine chooseCommand(@NotNull final Project project) {
+    final String androidHome = IntelliJAndroidSdk.chooseAndroidHome(project, false);
+
+    // Use daemon script if this is a bazel project.
+    final Workspace workspace = WorkspaceCache.getInstance(project).get();
+    if (workspace != null) {
+      final String script = workspace.getDaemonScript();
+      if (script != null) {
+        return createCommand(workspace.getRoot().getPath(), script, ImmutableList.of(), androidHome);
+      }
+
+    }
+
+    // Otherwise, use the Flutter SDK.
+    final FlutterSdk sdk = FlutterSdk.getFlutterSdk(project);
```

Comment:
what happens if the flutter sdk changes once this is already initialized?

Reply:
At this point nothing - the server will keep running and presumably there aren't many cases where users need a different version of devtools while working on a project. Are you thinking of some situations where this would be a problem?

Reply:
if you changed your flutter sdk the version of the devtools server much change so you want to make sure you have the one from the new sdk not the old one.

Reply:
Hmm... I'm not sure how to manage this situation. Do we have examples elsewhere of checking whether the flutter sdk has changed?

I'm also not sure about whether users change their flutter SDK frequently enough to monitor for this? My guess was that SDK would change if users update their flutter version or if they have multiple flutter downloads, but it would also surprise me that the implementation of DevTools would change significantly enough between those versions that a user would need the newest DevTools while in the middle of working on a project. Sounds like there may be some cases I'm not thinking of?